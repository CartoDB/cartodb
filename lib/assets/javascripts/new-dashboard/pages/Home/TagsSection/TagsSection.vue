<template>
  <section class="tags-section is-bgSoftBlue">
    <div class="container">
      <SectionTitle class="grid-cell">
        <template slot="icon">
          <img src="../../../assets/icons/section-title/tags.svg">
        </template>

        <template slot="title">{{ $t('HomePage.TagsSection.title') }}</template>

        <template slot="actionButton">
          <button class="button button--small is-primary button--ghost button--last" @click="goToRecentSection">
            <img svg-inline src="../../../assets/icons/sections/tags/recent_content.svg" class="tags__action"/>
            {{ $t('HomePage.TagsSection.viewRecentContentAction') }}
          </button>
        </template>
      </SectionTitle>

      <LoadingState v-if="isFetching" :text="$t('HomePage.TagsSection.loading')" class="loading"></LoadingState>

      <EmptyState
        :text="$t('HomePage.TagsSection.emptyState.title')"
        :subtitle="$t('HomePage.TagsSection.emptyState.subtitle')"
        v-if="isEmptyState">
      </EmptyState>

      <ul class="grid tags" v-if="hasTags">
        <li class="card grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile tag" v-for="tag in tags" :key="tag.tag">
          <TagCard :tag="tag" />
        </li>
      </ul>

      <Pagination class="pagination-element" v-if="shouldShowPagination" :page=currentPage :numPages=numPages @pageChange="goToPage"></Pagination>
    </div>
  </section>
</template>

<script>
import SectionTitle from 'new-dashboard/components/SectionTitle';
import TagCard from 'new-dashboard/components/Tag/TagCard';
import Pagination from 'new-dashboard/components/Pagination';
import EmptyState from 'new-dashboard/components/States/EmptyState';
import LoadingState from 'new-dashboard/components/States/LoadingState';

export default {
  name: 'TagsSection',
  components: {
    SectionTitle,
    TagCard,
    Pagination,
    EmptyState,
    LoadingState
  },
  mounted () {
    const page = parseInt(this.$route.query.sectionPage);
    this.goToPage(page);
  },
  data () {
    return {
      isFetching: false,
      tags: [],
      currentPage: 1,
      perPage: 6,
      totalTags: 0
    };
  },
  computed: {
    hasTags () {
      return !this.isFetching && this.totalTags > 0;
    },
    isEmptyState () {
      return !this.isFetching && this.totalTags <= 0;
    },
    numPages () {
      return Math.ceil(this.totalTags / this.perPage);
    },
    shouldShowPagination () {
      return this.numPages > 1;
    }
  },
  methods: {
    goToRecentSection () {
      const section = 'RecentSection';

      this.setURLParams({ section, sectionPage: 1 });
      this.$emit('sectionChange', section);
    },
    getTags (options = {}) {
      this.isFetching = true;

      options.perPage = this.perPage;

      this.$store.state.client.getTags(options,
        (err, _, tags) => {
          this.isFetching = false;

          if (err) {
            return;
          }

          this.tags = tags.result;
          this.totalTags = tags.total;
          this.currentPage = options.page;
        }
      );
    },
    goToPage (page) {
      this.setURLParams({ sectionPage: page });
      this.getTags({ page });
    },
    setURLParams ({ section = 'TagsSection', sectionPage = 1 }) {
      this.$router.push({ query: { section, sectionPage } });
    }
  }
};
</script>

<style lang="scss" scoped>
.tags {
  margin-bottom: 10px;
}

.tag {
  margin-bottom: 24px;
}

.tags__action {
  margin-right: 12px;
}

.loading {
  margin: 154px 0 85px;
}
</style>
