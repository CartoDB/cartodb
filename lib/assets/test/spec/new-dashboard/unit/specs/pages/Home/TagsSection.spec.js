import { shallowMount } from '@vue/test-utils';
import TagsSection from 'new-dashboard/pages/Home/TagsSection/TagsSection';

const tags = [
  { tag: 'Fake tag', maps: 1, datasets: 0 }
];

describe('TagsSection.vue', () => {
  let tagsSectionComponent, pushSpy;

  beforeEach(() => {
    pushSpy = jest.fn();
    tagsSectionComponent = createTagsSectionComponent({
      mocks: {
        $router: {
          push: pushSpy
        }
      }
    });
  });

  describe('Mounted', () => {
    it('should call goToPage with page number coming from URL if any', () => {
      const URLPage = 3;
      const mockComponent = createTagsSectionComponent({
        mocks: {
          $route: {
            query: { sectionPage: URLPage }
          },
          $router: {
            push: pushSpy
          }
        }
      });

      expect(mockComponent.vm.currentPage).toBe(URLPage);
    });
  });

  describe('Render', () => {
    it('should render properly', () => {
      tagsSectionComponent.setData({
        tags,
        isFetching: false,
        totalTags: 20
      });

      tagsSectionComponent.vm.$forceUpdate();

      expect(tagsSectionComponent).toMatchSnapshot();
    });

    it('should render empty state', () => {
      tagsSectionComponent.setData({
        tags: [],
        isFetching: false,
        totalTags: 0
      });

      tagsSectionComponent.vm.$forceUpdate();

      expect(tagsSectionComponent).toMatchSnapshot();
    });

    it('should render loading state', () => {
      tagsSectionComponent.setData({
        isFetching: true
      });

      tagsSectionComponent.vm.$forceUpdate();

      expect(tagsSectionComponent).toMatchSnapshot();
    });
  });

  describe('Computed', () => {
    it('numPages: should return the number of pages based on the number of tags per page', () => {
      tagsSectionComponent.setData({
        totalTags: 12
      });

      expect(tagsSectionComponent.vm.numPages).toBe(2);
    });

    it('shouldShowPagination: should return true if numPages is greater than 1', () => {
      tagsSectionComponent.setData({
        totalTags: 12
      });

      expect(tagsSectionComponent.vm.shouldShowPagination).toBe(true);
    });
  });

  describe('Methods', () => {
    it("goToRecentSection: should emit 'sectionChange' with value 'RecentSection' and change URL parameters", () => {
      const setURLParams = jest.fn();
      tagsSectionComponent.setMethods({ setURLParams });

      tagsSectionComponent.vm.goToRecentSection();

      expect(setURLParams).toHaveBeenCalledWith({ section: 'RecentSection', sectionPage: 1 });
      expect(tagsSectionComponent.emitted('sectionChange')).toBeTruthy();
      expect(tagsSectionComponent.emitted('sectionChange')[0][0]).toBe('RecentSection');
    });

    it('getTags: should call getTags and setURLParams', () => {
      const page = 1;
      tagsSectionComponent.vm.getTags({ page });

      expect(tagsSectionComponent.vm.tags).toBe(tags);
      expect(tagsSectionComponent.vm.totalTags).toBe(1);
      expect(tagsSectionComponent.vm.currentPage).toBe(1);
    });

    it('goToPage: should call getTags and setURLParams', () => {
      const page = 2;
      const setURLParams = jest.fn();
      const getTags = jest.fn();
      tagsSectionComponent.setMethods({ setURLParams, getTags });

      tagsSectionComponent.vm.goToPage(page);

      expect(setURLParams).toHaveBeenCalledWith({ sectionPage: page });
      expect(getTags).toHaveBeenCalledWith({ page });
    });

    it('setURLParams: should call router.push with new query parameters', () => {
      const section = 'TagsSection';
      const sectionPage = 2;

      tagsSectionComponent.vm.goToPage(sectionPage);

      expect(pushSpy).toHaveBeenCalledWith({ query: { section, sectionPage } });
    });
  });
});

function createTagsSectionComponent (options = {}) {
  return shallowMount(TagsSection, {
    mocks: {
      $t: key => key,
      $tc: key => key,
      $route: {
        query: { sectionPage: 1 }
      },
      $store: {
        state: {
          client: {
            getTags: (_, callback) => callback(null, null, { result: tags, total: 1 })
          }
        }
      },
      ...options.mocks
    }
  });
}
