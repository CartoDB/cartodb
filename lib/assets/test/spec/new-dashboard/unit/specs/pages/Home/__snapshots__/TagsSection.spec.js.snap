// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsSection.vue Render should render empty state 1`] = `
<section class="tags-section is-bgSoftBlue">
  <div class="container">
    <sectiontitle-stub class="grid-cell"><template></template> <template></template> <template></template></sectiontitle-stub>
    <!---->
    <emptystate-stub text="HomePage.TagsSection.emptyState.title" subtitle="HomePage.TagsSection.emptyState.subtitle"></emptystate-stub>
    <!---->
    <!---->
  </div>
</section>
`;

exports[`TagsSection.vue Render should render loading state 1`] = `
<section class="tags-section is-bgSoftBlue">
  <div class="container">
    <sectiontitle-stub class="grid-cell"><template></template> <template></template> <template></template></sectiontitle-stub>
    <loadingstate-stub text="HomePage.TagsSection.loading" class="loading"></loadingstate-stub>
    <!---->
    <!---->
    <!---->
  </div>
</section>
`;

exports[`TagsSection.vue Render should render properly 1`] = `
<section class="tags-section is-bgSoftBlue">
  <div class="container">
    <sectiontitle-stub class="grid-cell"><template></template> <template></template> <template></template></sectiontitle-stub>
    <!---->
    <!---->
    <ul class="grid tags">
      <li class="card grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile tag">
        <tagcard-stub tag="[object Object]"></tagcard-stub>
      </li>
    </ul>
    <pagination-stub page="1" numpages="4" class="pagination-element"></pagination-stub>
  </div>
</section>
`;
